service/src/main/java/ai/chronon/service/handlers/FeaturesHandler.java

```diff
@@ -0,0 +1,124 @@
+ * Some notes on this:
+ * We currently support bulkGet lookups against a single groupBy / join. Attempts to lookup n different GroupBys / Joins
+ * need to be split up into n different requests.
+ * A given bulkGet request might result in some successful lookups and some failed ones. We return a 4xx or 5xx response
+ * if the overall request fails (e.g. we're not able to parse the input json, Future failure due to Api returning an error)
+ * Individual failure responses will be marked as 'Failed' however the overall response status code will be successful (200)
+ * The response list maintains the same order as the incoming request list.
+ * As an example:
+ * { results: [ {"status": "Success", "features": ...}, {"status": "Failure", "error": ...} ] }
+ */
+public class FeaturesHandler implements Handler<RoutingContext> {
+
+    public enum EntityType {
+        GroupBy,
+        Join
+    }
+
+    private static final Logger logger = LoggerFactory.getLogger(FeaturesHandler.class);
+    private static final ObjectMapper objectMapper = new ObjectMapper();
+
+    private final EntityType entityType;
+    private final JavaFetcher fetcher;
+
+    public FeaturesHandler(EntityType entityType, JavaFetcher fetcher) {
+        this.entityType = entityType;
+        this.fetcher = fetcher;
+    }
+
+    @Override
+    public void handle(RoutingContext ctx) {
+        String entityName = ctx.pathParam("name");
+        logger.info("Retrieving {} - {}", entityType.name(), entityName);
+        JTry<List<JavaRequest>> maybeRequest = parseJavaRequest(entityName, ctx.body());
+        if (! maybeRequest.isSuccess()) {
+            logger.error("Unable to parse request body", maybeRequest.getException());
+            List<String> errorMessages = Collections.singletonList(maybeRequest.getException().getMessage());
+            ctx.response()
+                    .setStatusCode(400)
+                    .putHeader("content-type", "application/json")
+                    .end(new JsonObject().put("errors", errorMessages).encode());
+            return;
+        }
+
+        List<JavaRequest> requests = maybeRequest.getValue();
+        CompletableFuture<List<JavaResponse>> resultsJavaFuture =
+                entityType.equals(EntityType.GroupBy) ? fetcher.fetchGroupBys(requests) : fetcher.fetchJoin(requests);
+        // wrap the Java future we get in a Vert.x Future to not block the worker thread
+        Future<List<JTry<Map<String, Object>>>> maybeFeatureResponses =
+                Future.fromCompletionStage(resultsJavaFuture)
+                      .map(result -> result.stream().map(FeaturesHandler::responseToMap)
+                      .collect(Collectors.toList()));
+
+        maybeFeatureResponses.onSuccess(resultList -> {
+            // as this is a bulkGet request, we might have some successful and some failed responses
+            // we return the responses in the same order as they come in and mark them as successful / failed based
```

Comment:
Is order the only way of knowing what response maps to what request in the bulkget call? (still figuring out what the requests and responses look like)

Reply:
Yeah the order matches exactly. Similar pattern to other bulkget apis out there (e.g. [fb graph bulkget](https://developers.facebook.com/docs/graph-api/batch-requests) ).
This looks like:
Request - /v1/features/join/quickstart/training_set.v2
```
[{"user_id": "5"}, {"user_id": "6"}, {"user_id": "7"}]
```

Response:
```
{
  "results": [
      {"status": "Success", "features": { ...},
      {"status": "Success", "features": { ...},
      {"status": "Failure", "error": "Some exception",
  ]
}
```

Reply:
(I'd imagine calls would most often just be for a single element. I've wired it up as a bulkGet primarily cause the fetcher api is a bulk api and instead of sticking to a list of one it was easier to have things set up this way)

Reply:
Added a section to the README showing an example of this.